### ISSUE TITLE
[SDL2 & SDL3] Potential subsystem reference miscount from initialized "dependencies"

### ISSUE THREAD

**Issue by ivan-mogilko** [libsdl-org/SDL#8381]
**NOTE:** I have not looked too close, but *possibly* this may be also an issue in SDL3.

I was pointed out to this issue by another person (@ericoporto). Apparently SDL2 can end up with extra or not enough reference count for subsystems, which may result in either subsystems never removed (until ending SDL_Quit() call) or removed **too early**, maybe to a user's surprise.

This may happen when user calls SDL_InitSubSystem for subsystems that have dependencies, for example: Controller requires Joystick, several systems require Events, and so forth. As may be seen in this code here, SDL2 adds dependencies to initialization, which also leads to incrementing their reference counts:
https://github.com/libsdl-org/SDL/blob/32587bd7de23c057cf791e1b93e74885688c3e56/src/SDL.c#L190-L193

This dependency check is mirrored in SDL_QuitSubSystem, leading to decrementing the reference count of the "dependencies":
https://github.com/libsdl-org/SDL/blob/32587bd7de23c057cf791e1b93e74885688c3e56/src/SDL.c#L415-L417

So how may this be an issue?

Let's investigate two code examples.

Example 1:
```
// Init at program start one by one, maybe I do a step by step init reading from config, or something
SDL_InitSubSystem(SDL_INIT_VIDEO); // +1 EVENTS ref count
SDL_InitSubSystem(SDL_INIT_AUDIO); // +1 EVENTS ref count
SDL_InitSubSystem(SDL_INIT_JOYSTICKS); // +1 EVENTS ref count

// Uninit later, all at once
SDL_QuitSubSystem(SDL_INIT_VIDEO | SDL_INIT_AUDIO | SDL_INIT_JOYSTICKS); // -1 EVENTS ref count

// +2 EVENTS ref count remaining
```

Here, from a user's perspective, everything is fine, as the number of subsystems inits and uninits match. What user does not realize is that the number of dependencies inits and uninits does **not** match, leaving "dangling" references.

But this case is not as dangerous, because SDL_Quit will clear them anyway. However, there may be an opposite scenario

Example 2:
```
// Init at program start
SDL_InitSubSystem(SDL_INIT_VIDEO | SDL_INIT_AUDIO | SDL_INIT_JOYSTICKS); // +1 EVENTS ref count

// Uninit later, maybe i don't want audio for some reason
SDL_QuitSubSystem(SDL_INIT_AUDIO); // -1 EVENTS ref count

// EVENTS subsystem was removed at this point
```

Here EVENTS subsystem will be removed while there are still subsystems that supposedly require it.
This may be proved by calling `SDL_WasInit(SDL_INIT_EVENTS)` in between these steps.

---

I gave this a bit of thought, and the first solution that comes to mind is to increment dependency's refcounts *by the number of dependents*.

In other words, in case of:
```
SDL_InitSubSystem(SDL_INIT_VIDEO | SDL_INIT_AUDIO | SDL_INIT_JOYSTICKS);
```
there should be +3 EVENTS ref counts added. And same mirrored in SDL_QuitSubSystem.

But I don't know how compatible this is with the SDL's design, because user may also call SDL_QuitSubSystem(SDL_INIT_EVENTS) and probably expect to have it removed at once? Another approach may be to keep separate record of "internal" ref counts, but that will increase the code complexity.

**Comment by slouken** [comment]
Yes, this is a really good point. Would you like to submit a PR with a change to fix this?

**Comment by slouken** [comment]
Please target the main branch first. Thanks!
